File: assignments/zero-truncated-poisson/ztp.R
# Mikołaj Nowicki rozwiazanie w R

## z wykorzystaniem pakietu stats4 
library(stats4)
# install.packages("countreg", repos="http://R-Forge.R-project.org")
library(countreg)

##Przyjmijmy, że lambda to 5
set.seed(123)
N <- 100
wartosci <- rztpois(N, 5)

funkcja <- function(lambda) {
    wrt = logb(lambda, exp(1)) * sum(wartosci) - N * logb((exp(lambda) - 1), exp(1)) - 
      sum(logb(factorial(wartosci)), exp(1))
    - wrt
}

res <- mle(funkcja, start = list(lambda = 4.5), method = "L-BFGS-B")
summary(res)

File: assignments/zero-truncated-poisson/ztp.py
# Mikołaj Nowicki rozwiazanie w pythonie

import random
import math
from scipy.optimize import minimize
lambd=5
N=1000
lista=list()
for i in range(N):
  k = 1
  t = math.exp(-lambd)/(1-math.exp(-lambd))*lambd
  s = t
  u=random.random()
  while s<u:
    k=k+1
    t=t*lambd/k
    s=s+t
  lista.append(k)

wartosci=sum(lista)

def funkcja(x):
  wrt=math.log(x)*wartosci-N*math.log((math.exp(x)-1)) 
  return -wrt

res = minimize(funkcja,4.5)
print(res)
print(lista)
